discovery-provider/src/challenges/challenge_event_bus.py

```diff
@@ -40,21 +43,54 @@ def get_manager(self, challenge_id: str) -> ChallengeManager:
         """Gets a manager for a given challenge_id"""
         return self._managers[challenge_id]
 
+    @contextmanager
+    def use_scoped_dispatch_queue(self):
+        """Makes the bus only dispatch the events once out of the new scope created with 'with'"""
+        try:
+            self._use_in_memory_queue = True
+            yield self._in_memory_queue
+        finally:
+            logger.info(
+                f"ChallengeEventBus: Flushing {len(self._in_memory_queue)} events from in-memory queue"
+            )
+            self._use_in_memory_queue = False
+            for event in self._in_memory_queue:
+                self.dispatch(
+                    event["event"],
+                    event["block_number"],
+                    event["user_id"],
+                    event["extra"] if event["extra"] else {},
+                )
+            self._in_memory_queue.clear()
+
     def dispatch(
         self,
-        session: Session,
         event: str,
         block_number: int,
         user_id: int,
         extra: Dict = {},
     ):
-        """Dispatches an event + block_number + user_id to Redis queue"""
-        try:
-            event_json = self._event_to_json(event, block_number, user_id, extra)
-            logger.info(f"ChallengeEventBus: dispatch {event_json}")
-            self._redis.rpush(REDIS_QUEUE_PREFIX, event_json)
-        except Exception as e:
-            logger.warning(f"ChallengeEventBus: error enqueuing to Redis: {e}")
+        """Dispatches an event + block_number + user_id to Redis queue.
+
+        Note: May defer the dispatch if using an in-memory queue (see: use_scoped_dispatch_queue)
+        """
+        if self._use_in_memory_queue:
+            logger.info("ChallengeEventBus: deferring dispatch - using in-memory queue")
```

Comment:
this is gonna be a really noisy log since we dispatch an event for every play eventually

Reply:
maybe debug log?